### PR TITLE
Improve solver error reporting

### LINKED ISSUE #5086: No matching package when using ^ with pre-1.0 versions
With the following composer.json:

    {
        "name": "test/test",
        "authors": [
            {
                "name": "andig",
                "email": "test@test.de"
            }
        ],
        "require": {
            "react/socket-client": "^0.4",
            "react/http": "dev-master"
        }
    }

When I run this command:

```
composer install
composer require -vvv php-pm/php-pm:dev-master
```

I get this output:

	pi@prodpi ~/test2 $ composer require -vvv php-pm/php-pm:dev-master
	Reading ./composer.json
	Loading config file /home/pi/.composer/config.json
	Loading config file /home/pi/.composer/auth.json
	Loading config file ./composer.json
	Checking CA file /etc/ssl/certs/ca-certificates.crt
	Executing command (/home/pi/test2): git branch --no-color --no-abbrev -v
	Failed to initialize global composer: Composer could not find the config file: /home/pi/.composer/composer.json
	To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
	Reading /home/pi/test2/vendor/composer/installed.json
	./composer.json has been updated
	Reading ./composer.json
	Loading config file /home/pi/.composer/config.json
	Loading config file /home/pi/.composer/auth.json
	Loading config file ./composer.json
	Executing command (/home/pi/test2): git branch --no-color --no-abbrev -v
	Failed to initialize global composer: Composer could not find the config file: /home/pi/.composer/composer.json
	To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
	Reading /home/pi/test2/vendor/composer/installed.json
	Reading ./composer.lock
	Loading composer repositories with package information
	Downloading https://packagist.org/packages.json
	Writing /home/pi/.composer/cache/repo/https---packagist.org/packages.json into cache
	Updating dependencies (including require-dev)
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2013.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2014.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2015.json from cache
	Downloading http://packagist.org/p/provider-2015-04%24613b05fce10207a1c94baf39ecbf968560f853292b6b68f34b96fa4696658f86.json
	Writing /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2015-04.json into cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2015-07.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2015-10.json from cache
	Downloading http://packagist.org/p/provider-2016-01%24c3a38e6d71d9ee37a49e4ea3fed55a08830cba62142c3e8192ea9f86eb2fc8db.json
	Writing /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2016-01.json into cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-archived.json from cache
	Downloading http://packagist.org/p/provider-latest%240399eaaa23be5dd3ad83bee459e79fd672e77cc88263af05bc355515b21a6848.json
	Writing /home/pi/.composer/cache/repo/https---packagist.org/p-provider-latest.json into cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$promise.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-evenement$evenement.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$stream.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$event-loop.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$socket.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$cache.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$dns.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$socket-client.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-psr$http-message.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-guzzlehttp$psr7.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$http.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-php-pm$php-pm.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-mkraemer$react-pcntl.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-symfony$console.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-symfony$process.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-symfony$debug.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-monolog$monolog.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-symfony$polyfill-mbstring.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-psr$log.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-symfony$http-kernel.json from cache
	Your requirements could not be resolved to an installable set of packages.

	  Problem 1
	    - Installation request for php-pm/php-pm dev-master -> satisfiable by php-pm/php-pm[dev-master].
	    - php-pm/php-pm dev-master requires react/socket-client ^0.5.0 -> no matching package found.

	Potential causes:
	 - A typo in the package name
	 - The package is not available in a stable-enough version according to your minimum-stability setting
	   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.

	Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.

	Installation failed, reverting ./composer.json to its original content.

**And I expected this to happen:**

Instead of installing, I see an error `react/socket-client ^0.5.0 -> no matching package found`. However, simply running `composer require react/socket-client:^0.5` works fine.

As ^ operator forbids loading 0.5 if 0.4 is already installed I was expecting a version constraint mismatch, bot a "not found". As 0.5.0 for this package had only just been released it kept chasing me up the wrong tree of caches etc.

Wrong expectation on composer error messages?

### PR DESCRIPTION
**Pull request by Seldaek**
Fixes #5086, #2575 and #2661

Using this:

```json
{
    "require": {
        "react/socket-client": "^0.4",
        "react/http": "dev-master",
        "seld/jsonlint": "dev-nope",
        "foo/madeup": "dev-master",
        "relution-developments/carbon-php": "*",
        "php-pm/php-pm": "dev-master"
    }
}
```

We currently get:

```
  Problem 1
    - The requested package seld/jsonlint could not be found in any version, there may be a typo in the package name.
  Problem 2
    - The requested package foo/madeup could not be found in any version, there may be a typo in the package name.
  Problem 3
    - The requested package relution-developments/carbon-php could not be found in any version, there may be a typo in the package name.
  Problem 4
    - Installation request for php-pm/php-pm dev-master -> satisfiable by php-pm/php-pm[dev-master].
    - php-pm/php-pm dev-master requires react/socket-client ^0.5.0 -> no matching package found.
```

Which unhelpfully tells users that packages probably don't exist in problem 1, 3 and 4 (only 2 is correct).

With the patch, we get this:

```
  Problem 1
    - The requested package seld/jsonlint dev-nope exists as seld/jsonlint[1.0.0, 1.0.1, 1.1.0, 1.1.1, 1.1.2, 1.2.0, 1.3.0, 1.3.1, 1.4.0, dev-master] but those are rejected by your constraint.
  Problem 2
    - The requested package foo/madeup could not be found in any version, there may be a typo in the package name.
  Problem 3
    - The requested package relution-developments/carbon-php * is satisfiable by relution-developments/carbon-php[dev-master] but those are rejected by your minimum-stability.
  Problem 4
    - Installation request for php-pm/php-pm dev-master -> satisfiable by php-pm/php-pm[dev-master].
    - php-pm/php-pm dev-master requires react/socket-client ^0.5.0 -> satisfiable by react/socket-client[v0.5.0] but these conflict with your requirements or minimum-stability
```

### CODE REVIEW

**Comment by Seldaek** [comment]
@naderman if you have 2min to do a cursory check here make sure I didn't corner us for future improvements it'd be good. But otherwise I think it's fine to merge.

File: src/Composer/DependencyResolver/Pool.php [hunk 1]
```diff
@@ -176,27 +176,32 @@ public function count()
      *                                            packages must match or null to return all
      * @param  bool                $mustMatchName Whether the name of returned packages
      *                                            must match the given name
+     * @param  bool                $bypassFilters If enabled, filterRequires and stability matching is ignored
      * @return PackageInterface[]  A set of packages
      */
-    public function whatProvides($name, ConstraintInterface $constraint = null, $mustMatchName = false)
+    public function whatProvides($name, ConstraintInterface $constraint = null, $mustMatchName = false, $bypassFilters = false)
     {
+        if ($bypassFilters === true) {
```
Comment (stof) [inline src/Composer/DependencyResolver/Pool.php]: ``=== true`` is useless, as the variable is a boolean.

File: src/Composer/DependencyResolver/Problem.php [hunk 2]
```diff
@@ -229,4 +233,21 @@ protected function constraintToText($constraint)
     {
         return ($constraint) ? ' '.$constraint->getPrettyString() : '';
     }
+
+    protected function formatPackagesUnique($pool, array $packages)
```
Comment (stof) [inline src/Composer/DependencyResolver/Problem.php]: should be private IMO

File: src/Composer/Repository/ComposerRepository.php [hunk 3]
```diff
@@ -396,7 +401,18 @@ public function whatProvides(Pool $pool, $name)
             }
         }
 
-        return $this->providers[$name];
+        $result = $this->providers[$name];
+
+        // clean up the cache because otherwise using this puts the repo in an inconsistent state with a polluted unfiltered cache
+        // which is likely not an issue but might cause hard to track behaviors depending on how the repo is used
+        if ($bypassFilters) {
```
Comment (stof) [inline src/Composer/Repository/ComposerRepository.php]: shouldn't we restore the previous cache in this case, if possible (and existing of course) ?

**Comment by stof** [comment]
would be great to have tests covering this

File: src/Composer/Repository/ComposerRepository.php (hunk 3 above)
Comment (Seldaek) [inline src/Composer/Repository/ComposerRepository.php]: I don't think it matters really. It's gonna be a lot of code for something that's kinda useless, as once we are done outputting a solving error we essentially abort.

**Comment by Seldaek** [comment]
Added tests & updated code.

**Comment by naderman** [comment]
@Seldaek for the record this is fine with my other WIP PR

**Comment by Emanuel937** [comment]
I'm getting the some error

**Pull request merged**